### PR TITLE
Bug 1278767 -  (Re)start remote-control service when time is changed or networks is connected

### PR DESCRIPTION
**Pull request by ChunMinChang**

### CODE REVIEW

**Comment by schien** [comment]
We use anonymous function to create a function-local scope for all the remote control variables / functions. This prevents name conflict and function misused in other JS script.
If there is no specific reason, please keep the same structure and add only necessary code for monitoring timezone change.

File: b2g/chrome/content/remote_control.js [hunk 1]
```diff
@@ -72,5 +72,5 @@ const kPrefPresentationDiscoverable = "dom.presentation.discoverable";
     Services.obs.addObserver(networkStatusMonitor, "network:offline-status-changed", false);
   }
 
-  setupRemoteControl();
+  window.onmoztimechange = setupRemoteControl;
```
Comment (schien) [inline b2g/chrome/content/remote_control.js]: This patch is too specific to the boot sequence of Nexus player. What we should do here is adding a logic of:
1. RemoteControlService.stop() if service is running.
2. delete the previous generated key
3. RemoteControlService.start() if necessary.

File: b2g/chrome/content/remote_control.js [hunk 2]
```diff
@@ -20,57 +20,99 @@ const kPrefPresentationDiscoverable = "dom.presentation.discoverable";
...
-  Services.prefs.addObserver(kPrefPresentationDiscoverable, setupRemoteControl, false);
+  // Monitor the network status and the time change
+  var statusMonitor = {
+    init: function sm_init() {
+      debug("init");
+
+      Services.prefs.addObserver(kPrefPresentationDiscoverable, this, false);
+      Services.obs.addObserver(this, "network-active-changed", false);
+      Services.obs.addObserver(this, "network:offline-status-changed", false);
+      window.onmoztimechange = this.onTimeChange;
```
Comment (schien) [inline b2g/chrome/content/remote_control.js]: Please use addEventListener, otherwise you might accidentally remove other people's timechnage listener.

File: b2g/components/RemoteControlService.jsm [hunk 3]
```diff
@@ -325,6 +325,16 @@ this.RemoteControlService = {
...
   },
 
+  // Refresh the certificate when the time is updated
+  // The getOrCreateCert will validate the certificate itself.
+  // It is invalid when (current time - cert's expired day) < one day,
+  // or the (current time < cert's started day), otherwise we can use
+  onTimeChange: function(aTimestamp) {
+    DEBUG && debug("time is updated to: " + new Date(aTimestamp).toString());
+    this.stop();
+    this.start();
```
Comment (schien) [inline b2g/components/RemoteControlService.jsm]: This looks wrong to me because the server might not be start at the beginning.

File: b2g/chrome/content/remote_control.js [hunk 4]
```diff
@@ -20,57 +20,99 @@ const kPrefPresentationDiscoverable = "dom.presentation.discoverable";
...
 
-  Services.prefs.addObserver(kPrefPresentationDiscoverable, setupRemoteControl, false);
+  // Monitor the network status and the time change
+  var statusMonitor = {
+    init: function sm_init() {
+      debug("init");
+
+      Services.prefs.addObserver(kPrefPresentationDiscoverable, this, false);
+      Services.obs.addObserver(this, "network-active-changed", false);
+      Services.obs.addObserver(this, "network:offline-status-changed", false);
```
Comment (schien) [inline b2g/chrome/content/remote_control.js]: Keep the |if (Ci.nsINetworkManager) {| logic.

File: b2g/chrome/content/remote_control.js [hunk 5]
```diff
@@ -20,57 +20,99 @@ const kPrefPresentationDiscoverable = "dom.presentation.discoverable";
...
+      remoteControlScope.RemoteControlService.stop();
+    },
+
+    get discoverable() {
+      return Services.prefs.getBoolPref(kPrefPresentationDiscoverable);
+    },
+  };
 
-  setupRemoteControl();
+  statusMonitor.init();
```
Comment (schien) [inline b2g/chrome/content/remote_control.js:133]: |statusMonitor.init| only do observer registration but not perform the service start during boot-up. Is this change intentional?

Comment (ChunMinChang) [inline b2g/chrome/content/remote_control.js:133]: I mentioned this issue on [0]. If remote-control service is started on boot-up, and it's restarted on receiving time-change event, then we will get the following sequence:
1. flash
   1. Init time is 2016/xx/xx
   2. start service -> getOrCreateCert -> Get cert fail -> create cert 'A' started from 2016... 
   3. onTimeChange -> time is 2016/xx/xx -> getOrCreateCert -> Get cert 'A' -> validate successfully
2. first boot
   1. Init time is 2000/xx/xx
   2. start service -> getOrCreateCert -> Get cert 'A' -> validate fail -> create cert 'B' started from 2000...
   3. onTimeChange -> time is 2016/xx/xx -> getOrCreateCert -> Get cert 'B' -> validate fail -> create cert 'B_1' started from 2016...
3. second boot
   1. Init time is 2000/xx/xx
   2. start service -> getOrCreateCert -> Get cert 'B_1' -> validate fail -> create cert 'C' started from 2000...
   3. onTimeChange -> time is 2016/xx/xx -> getOrCreateCert  -> Get cert 'C' -> validate fail -> create cert 'C_1' started from 2016...
      ...
      ...
      ...
4. .....

When |getOrCreateCert|[1] is called for starting service, it will validate the certificate's period. If current time < cert's started day or (current time - cert's expired day) < 1 day, then it's invalid[2]. 

Thus, if the code here is

``` javascript
statusMonitor.init();

if (statusMonitor.startService()) {
  statusMonitor.stopService();
}
```

The certificate will be changed all the times.

[0] https://bugzilla.mozilla.org/show_bug.cgi?id=1278767#c24
[1] https://dxr.mozilla.org/mozilla-central/source/security/manager/ssl/nsILocalCertService.idl#30
[2] http://searchfox.org/mozilla-central/source/security/manager/ssl/LocalCertService.cpp#331

File: b2g/components/RemoteControlService.jsm (hunk 3 above)
Comment (ChunMinChang) [inline b2g/components/RemoteControlService.jsm]: If the time-change event is not fired when booting up, then the service will be started after the network is connected.

File: b2g/chrome/content/remote_control.js (hunk 5 above)
Comment (schien) [inline b2g/chrome/content/remote_control.js:133]: Thanks for the detailed explanation. I'm still worrying about if these event timing are specific to Nexus Player. Could you verify your patch with Mulet as well?

Comment (ChunMinChang) [inline b2g/chrome/content/remote_control.js:133]: If the time-change event is fired before network is connected, then `onTimeChange` will be the entry point for the service, otherwise, the service will be started upon network is alive, so I think it's not specific for nexus player. 

I'll check  on mulet and see what happens.

File: b2g/components/RemoteControlService.jsm (hunk 3 above)
Comment (schien) [inline b2g/components/RemoteControlService.jsm]: No, what I mean is Remote Control service might be pref off. We don't want to start the server even if ontimechange is received.

File: b2g/chrome/content/remote_control.js (hunk 5 above)
Comment (ChunMinChang) [inline b2g/chrome/content/remote_control.js:133]: In this version, the service would be started on booting up.

File: b2g/components/RemoteControlService.jsm [hunk 6]
```diff
@@ -381,84 +407,129 @@ this.RemoteControlService = {
...
-            aReject("Start TLSSocketServer fail");
-            return;
+      // Try toget certificate from x509cert DB first.
+      // If the certificate doesn't exist, then create one.
+      try {
+        let certDB = Cc["@mozilla.org/security/x509certdb;1"]
+                     .getService(Ci.nsIX509CertDB);
+        let cert = certDB.findCertByNickname(SERVICE_CERT);
+        DEBUG && debug("Get the existing certificate");
+        aResolve(cert);
```
Comment (schien) [inline b2g/components/RemoteControlService.jsm:467]: This approach means the self-signed certificate will not be renewed after expiry.

**Comment by schien** [comment]
This patch depends on a Nexus Player (or Firefox OS on AOSP) specific behavior, which is the system time is always wrong after reboot and there is always a moztimechange event afterward. My concern is that this behavior might not existed on the production environment and your solution here will break the production environment.

I suggest we keep only minimum workaround for Nexus Player and put it behind a flag. If the same issue happened in real product, we can simply remove the flag.

File: b2g/components/RemoteControlService.jsm (hunk 6 above)
Comment (ChunMinChang) [inline b2g/components/RemoteControlService.jsm:467]: Wrap the code depending on Gonk into the preprocessor directives now.

File: b2g/chrome/content/remote_control.js [hunk 7]
```diff
@@ -20,57 +20,113 @@ const kPrefPresentationDiscoverable = "dom.presentation.discoverable";
...
+      Services.prefs.addObserver(kPrefPresentationDiscoverable, this, false);
+
+#ifdef MOZ_WIDGET_GONK
+      window.addEventListener('moztimechange', this.onTimeChange);
+
+      if (Ci.nsINetworkManager) {
+        Services.obs.addObserver(this, "network-active-changed", false);
+        Services.obs.addObserver(this, "network:offline-status-changed", false);
+      }
+#endif
```
Comment (schien) [inline b2g/chrome/content/remote_control.js]: operation related to network status shouldn't be gonk-specific.

File: b2g/components/RemoteControlService.jsm (hunk 6 above)
Comment (schien) [inline b2g/components/RemoteControlService.jsm:467]: you can try following style, which will avoid the ugly |#else| block.

#ifdef MOZ_WIDGET_GONK
  try {
    let cert = certDB.findCertByNickname(...);
    aResolve(cert);
    return;
  } catch (e) {
    // cert cannot be found, fall through to create new one.
  }
#endif

certService.getOrCreate(...);

File: b2g/chrome/content/remote_control.js (hunk 7 above)
Comment (ChunMinChang) [inline b2g/chrome/content/remote_control.js]: Hmm....the [nsINetworkManager.idl](https://dxr.mozilla.org/mozilla-central/source/dom/system/gonk/nsINetworkManager.idl) is under `dom/system/gonk`[0]

Does partner's platform also has their own nsINetworkManager?

[0] https://dxr.mozilla.org/mozilla-central/source/dom/system/gonk/nsINetworkManager.idl
